src/include/ibverbs_wrapper.hpp

```diff
@@ -1,301 +1,107 @@
-      return ibv_reg_mr_lib(pd, addr, length, access);
-    }
-    return nullptr;
-  }
-
-  // Static method to register a dma-buf based memory region
-  static struct ibv_mr* ibv_reg_dmabuf_mr(struct ibv_pd* pd, uint64_t offset, size_t length, uint64_t iova, int fd,
-                                          int access) {
-    if (!initialized) initialize();
-    if (ibv_reg_dmabuf_mr_lib) {
-      return ibv_reg_dmabuf_mr_lib(pd, offset, length, iova, fd, access);
-    }
-    return nullptr;
-  }
-
-  // Static method to deregister a memory region
-  static int ibv_dereg_mr(struct ibv_mr* mr) {
-    if (!initialized) initialize();
-    if (ibv_dereg_mr_lib) {
-      return ibv_dereg_mr_lib(mr);
-    }
-    return -1;
-  }
-
-  // Static method to query a GID
-  static int ibv_query_gid(struct ibv_context* context, uint8_t port_num, int index, union ibv_gid* gid) {
-    if (!initialized) initialize();
-    if (ibv_query_gid_lib) {
-      return ibv_query_gid_lib(context, port_num, index, gid);
-    }
-    return -1;
-  }
-
-  // Static method to modify a queue pair
-  static int ibv_modify_qp(struct ibv_qp* qp, struct ibv_qp_attr* attr, int attr_mask) {
-    if (!initialized) initialize();
-    if (ibv_modify_qp_lib) {
-      return ibv_modify_qp_lib(qp, attr, attr_mask);
-    }
-    return -1;
-  }
-
-  // Static method to destroy a queue pair
-  static int ibv_destroy_qp(struct ibv_qp* qp) {
-    if (!initialized) initialize();
-    if (ibv_destroy_qp_lib) {
-      return ibv_destroy_qp_lib(qp);
-    }
-    return -1;
-  }
-
-  static inline int ibv_post_send(struct ibv_qp* qp, struct ibv_send_wr* wr, struct ibv_send_wr** bad_wr) {
-    if (!initialized) initialize();
+#define REGISTER_IBV_FUNC_WITH_NAME(name__, func__)                                          \
+  template <typename... Args>                                                                \
+  static auto(name__)(Args && ...args) {                                                     \
+    static_assert(sizeof(&::func__) > 0, #func__ " is expected be a function, not a macro"); \
+    static decltype(&::func__) impl = nullptr;                                               \
+    if (!impl) impl = reinterpret_cast<decltype(impl)>(IBVerbs::dlsym(#func__));             \
+    return impl(std::forward<Args>(args)...);                                                \
+  }
+
+#define REGISTER_IBV_FUNC(func__) REGISTER_IBV_FUNC_WITH_NAME(func__, func__)
+
+  ///
+  /// Usual cases where we can link the function directly
+  ///
+
+  REGISTER_IBV_FUNC(ibv_free_device_list)
+  REGISTER_IBV_FUNC(ibv_alloc_pd)
+  REGISTER_IBV_FUNC(ibv_dealloc_pd)
+  REGISTER_IBV_FUNC(ibv_open_device)
+  REGISTER_IBV_FUNC(ibv_close_device)
+  REGISTER_IBV_FUNC(ibv_query_device)
+  REGISTER_IBV_FUNC(ibv_create_cq)
+  REGISTER_IBV_FUNC(ibv_destroy_cq)
+  REGISTER_IBV_FUNC(ibv_create_qp)
+  REGISTER_IBV_FUNC(ibv_destroy_qp)
+  REGISTER_IBV_FUNC(ibv_modify_qp)
+  REGISTER_IBV_FUNC(ibv_reg_dmabuf_mr)
```

Comment:
Seems ibv_reg_dambuf_mr only avalibale for version 1.12. May need to checkout version here